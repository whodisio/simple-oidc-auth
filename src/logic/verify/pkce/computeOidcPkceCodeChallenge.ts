import { sha256 } from 'cross-sha256';

import { castBase64ToBase64Url } from '../../base64Url/castBase64ToBase64Url';

/**
 * computes the pkce code challenge from the pkce code verifier
 *
 * spec
 * ```
 *       code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
 * ```
 *
 * ref
 * - https://datatracker.ietf.org/doc/html/rfc7636#section-4.2
 * - https://stackoverflow.com/questions/63309409/creating-a-code-verifier-and-challenge-for-pkce-auth-on-spotify-api-in-reactjs
 */
export const computeOidcPkceCodeChallenge = ({
  verifier,
}: {
  verifier: string;
}): string =>
  castBase64ToBase64Url(new sha256().update(verifier).digest('base64'));
