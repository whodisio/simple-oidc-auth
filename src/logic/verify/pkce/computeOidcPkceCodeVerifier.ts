import { sha256 } from 'cross-sha256';

/**
 *
 * requirements
 * - sufficient entropy - âœ…, challenge uuid = a uuid, completely random -> increased to 64 char through sha256
 *
 * ref
 * - https://datatracker.ietf.org/doc/html/rfc7636#section-4.1
 */
export const computeOidcPkceCodeVerifier = ({
  oidcRequestUuid,
  userSessionUuid,
}: {
  /**
   * the primary key identifier of the request, persisted by your backend
   */
  oidcRequestUuid: string;

  /**
   * the primary key identifier of the current user's session, if any
   */
  userSessionUuid: string | null;
}): string =>
  [
    'opcv',
    new sha256()
      .update(
        JSON.stringify({
          purpose: 'oidcPkceCodeVerifier',
          oidcRequestUuid,
          userSessionUuid,
        }),
      )
      .digest('hex'),
  ].join('_');
