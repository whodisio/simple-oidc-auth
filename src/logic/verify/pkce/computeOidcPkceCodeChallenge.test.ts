import { computeOidcPkceCodeChallenge } from './computeOidcPkceCodeChallenge';
import { computeOidcPkceCodeVerifier } from './computeOidcPkceCodeVerifier';

describe('computeOidcPkceCodeChallenge', () => {
  it('should be deterministic', () => {
    const inputs = {
      oidcRequestUuid: '49cdec28-7a27-42f0-be33-dd3673255bde',
      userSessionUuid: null,
    };
    const verifier = computeOidcPkceCodeVerifier(inputs);
    const challenge = computeOidcPkceCodeChallenge({ verifier });
    expect(computeOidcPkceCodeChallenge({ verifier })).toEqual(challenge);
    expect(challenge).toMatchSnapshot(); // log an example
    expect(challenge.length).toEqual(43); // should always be 43
    expect(challenge).toEqual('ZSo1zPq28s1l5x11Z7aG9jG2ezcdBDeyHTAJBkSsf1Q'); // based on https://tonyxu-io.github.io/pkce-generator/
    expect(challenge).toEqual('ZSo1zPq28s1l5x11Z7aG9jG2ezcdBDeyHTAJBkSsf1Q'); // based on https://example-app.com/pkce
  });
});
