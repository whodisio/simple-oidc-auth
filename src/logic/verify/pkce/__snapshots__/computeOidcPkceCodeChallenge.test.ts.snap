// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeOidcPkceCodeChallenge should be deterministic 1`] = `"ZSo1zPq28s1l5x11Z7aG9jG2ezcdBDeyHTAJBkSsf1Q"`;
